### PR TITLE
feat: [IOBP-6] Addition of pre-error mapping IDPay post-auth screens

### PR DESCRIPTION
**Pull request by forrest57**
## Short description
Added success, canceled and generic error transaction results screens.
Screenshot 2023-06-12 at 15 43 04

## List of changes proposed in this pull request
- added required I18n keys
- added an xstate selector for the current error state
this selector also makes sure that the machine is in an error state 
- added previously listed components

## How to test
once again, `io-dev-server` is your friend here.
Navigate to a discount initiative's details screen, tap on the "make payment" CTA, and insert:
- 00000000 to have a generic error
- 00000001 to have a valid payload, from which you can:
        - cancel the transaction to have the `CANCELED` error screen
        - authorize the transaction to have the `SUCCESS` screen

### CODE REVIEW

File: ts/features/idpay/payment/screens/IDPayPaymentResultScreen.tsx [hunk 1]
```diff
@@ -1,34 +1,76 @@
...
+        </Body>
+      )}
+    </View>
   );
 
+  const content =
+    possibleFailureType !== undefined // also checks for failure state to be true
+      ? renderContent(possibleFailureType)
+      : renderContent("SUCCESS");
+
```
Comment (mastro993) [inline ts/features/idpay/payment/screens/IDPayPaymentResultScreen.tsx:70]:
```suggestion
   const failureOption = useSelector(machine, selectFailureOption);

    const renderContent = (resultType: PaymentResultType) => (
    <View style={resultScreenStyles.resultScreenContainer}>
      <Pictogram name={results[resultType].icon} size={120} />
      <VSpacer size={24} />
      <NewH4 style={resultScreenStyles.justifyTextCenter}>
        {results[resultType].title}
      </NewH4>
      <VSpacer size={16} />
      {results[resultType].subtitle && (
        <Body style={resultScreenStyles.justifyTextCenter}>
          {results[resultType].subtitle}
        </Body>
      )}
    </View>
  );

  const content = pipe(
    failureOption,
    O.fold(
      () => renderContent("SUCCESS"),
      failure => renderContent(failure)
    )
  );

```

What about this?

File: ts/features/idpay/payment/screens/IDPayPaymentResultScreen.tsx [hunk 2]
```diff
@@ -1,34 +1,76 @@
...
+    ["SUCCESS"]: {
+      title: I18n.t("idpay.payment.result.success.title"),
+      subtitle: I18n.t("idpay.payment.result.success.body"),
+      icon: "completed"
+    }
+  };
+
+  const renderContent = (resultType: PaymentResults) => (
+    <View style={resultScreenStyles.resultScreenContainer}>
+      <Pictogram name={results[resultType].icon} size={120} />
```
Comment (mastro993) [inline ts/features/idpay/payment/screens/IDPayPaymentResultScreen.tsx:52]:
```suggestion
      <Pictogram name={results[resultType].icon} size={80} />
```

**Review by mastro993 — CHANGES_REQUESTED** [review]

File: ts/features/idpay/payment/screens/IDPayPaymentResultScreen.tsx [hunk 3]
```diff
@@ -1,34 +1,76 @@
...
+      <VSpacer size={24} />
```
[the remark this answers is not in the document]
Comment (forrest57) [reply ts/features/idpay/payment/screens/IDPayPaymentResultScreen.tsx:53]: the spacing in the design is  inconsistent, ranging from 24 in the CANCELED and most other error cases to 16 in the success case, and about 30 in the GENERIC error case, so after initially testing with a spacing of 16 I came to the conclusion that 24 was what best fit design; do you believe that 16 is better in error cases as well? We could also dynamically render the spacing but I am not sure that could be the best idea

[the remark this answers is not in the document]
Comment (thisisjp) [reply ts/features/idpay/payment/screens/IDPayPaymentResultScreen.tsx:53]: this page will be redesigned soon 🥲 24 is fine

File: ts/features/idpay/payment/screens/IDPayPaymentResultScreen.tsx (hunk 2 above)
Comment (forrest57) [reply ts/features/idpay/payment/screens/IDPayPaymentResultScreen.tsx:52]: after a short discussion on Figma, 120 was approved as the uniform sizing for pictograms in this flow 👍

**Comment by forrest57** [comment]
@mastro993  changes adressed in https://github.com/pagopa/io-app/pull/4682/commits/70a864caa87b00121263329081622ab37be7c4b1

File: ts/features/idpay/payment/screens/IDPayPaymentResultScreen.tsx [hunk 4]
```diff
@@ -1,34 +1,80 @@
...
 import { usePaymentMachineService } from "../xstate/provider";
-import { selectIsFailure, selectTransactionData } from "../xstate/selectors";
+import { selectFailureOption } from "../xstate/selectors";
+
+type FailureContent = {
+  title: string;
+  subtitle?: string;
+  icon: IOPictograms;
+};
+type PaymentResultsType = PaymentFailure | "SUCCESS";
```
Comment (mastro993) [inline ts/features/idpay/payment/screens/IDPayPaymentResultScreen.tsx:25]:
```suggestion
type PaymentResultType = PaymentFailure | "SUCCESS";
```
Use the singular when defining types.

**Review by mastro993 — CHANGES_REQUESTED** [review]

File: ts/features/idpay/payment/screens/IDPayPaymentResultScreen.tsx (hunk 4 above)
Comment (forrest57) [reply ts/features/idpay/payment/screens/IDPayPaymentResultScreen.tsx]: adressed in https://github.com/pagopa/io-app/pull/4682/commits/3cff96b19bae16d42fd7ae3071da832a6a865220

**Review by mastro993 — APPROVED** [review]
LGTM

**Pull request merged**